README.md
# Check Operator for the phonenumber in cameroon

[![Latest Version on Packagist](https://img.shields.io/packagist/v/ gildastema/phonenumber.svg?style=flat-square)](https://packagist.org/packages/ gildastema/phonenumber)
[![Total Downloads](https://img.shields.io/packagist/dt/ gildastema/phonenumber.svg?style=flat-square)](https://packagist.org/packages/ gildastema/phonenumber)
![GitHub Actions](https://github.com/ gildastema/phonenumber/actions/workflows/main.yml/badge.svg)

Check Operator for the phonenumber in cameroon

## Installation

You can install the package via composer:

```bash
composer require  gildastema/phonenumber
```

## Usage

```php
    (new \Gildastema\Phonenumber\Phonenumber())->getOperator('691131446');
    (new \Gildastema\Phonenumber\Phonenumber())->getOperator('237691131446');
```

### Testing

```bash
composer test
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

### Security

If you discover any security related issues, please email gildastema3@gmail.com instead of using the issue tracker.

## Credits

-   [Gildas Tema](https://github.com/gildastema)
-   [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

## PHP Package Boilerplate

This package was generated using the [PHP Package Boilerplate](https://laravelpackageboilerplate.com) by [Beyond Code](http://beyondco.de/).


CHANGELOG.md
# Changelog

All notable changes to `phonenumber` will be documented in this file

## 1.0.0 - 201X-XX-XX

- initial release


tests/PhoneFakerTest.php
<?php

namespace Gildastema\Phonenumber\Tests;

use Gildastema\Phonenumber\PhoneFaker;
use Gildastema\Phonenumber\Phonenumber;
use PHPUnit\Framework\TestCase;

class PhoneFakerTest extends TestCase
{
    private PhoneFaker $phoneFaker;
    private Phonenumber  $phonenumber;

    protected function setUp(): void
    {
        parent::setUp(); // TODO: Change the autogenerated stub
        $this->phoneFaker = new PhoneFaker();
        $this->phonenumber = new Phonenumber();
    }

    /** @test  */
    public function orange()
    {
        $this->assertEquals(Phonenumber::ORANGE, $this->phonenumber->getOperator($this->phoneFaker->generate(Phonenumber::ORANGE)));
    }

    /** @test  */
    public function mtn()
    {
        $this->assertEquals(Phonenumber::MTN, $this->phonenumber->getOperator($this->phoneFaker->generate(Phonenumber::MTN)));
    }
}

tests/PhonenumberTest.php
<?php

namespace Gildastema\Phonenumber\Tests;

use Gildastema\Phonenumber\Phonenumber;
use PHPUnit\Framework\TestCase;

class PhonenumberTest extends TestCase
{
    private Phonenumber $phone;
    protected function setUp(): void
    {
        parent::setUp(); // TODO: Change the autogenerated stub
        $this->phone = new Phonenumber();
    }

    /** @test */
    public function orange()
    {
        $this->assertEquals(Phonenumber::ORANGE , $this->phone->getOperator('691131446') );
        $this->assertEquals(Phonenumber::ORANGE , $this->phone->getOperator('655236987') );
        $this->assertEquals(Phonenumber::ORANGE , $this->phone->getOperator('656236987') );
        $this->assertEquals(Phonenumber::ORANGE , $this->phone->getOperator('657236987') );
        $this->assertEquals(Phonenumber::ORANGE , $this->phone->getOperator('658236987') );
        $this->assertEquals(Phonenumber::ORANGE , $this->phone->getOperator('659236987') );
    }
    /** @test  */
    public function mtn()
    {
        $this->assertEquals(Phonenumber::MTN , $this->phone->getOperator('677609512') );
        $this->assertEquals(Phonenumber::MTN , $this->phone->getOperator('680609512') );
        $this->assertEquals(Phonenumber::MTN , $this->phone->getOperator('681609512') );
        $this->assertEquals(Phonenumber::MTN , $this->phone->getOperator('650236987') );
        $this->assertEquals(Phonenumber::MTN , $this->phone->getOperator('651236987') );
        $this->assertEquals(Phonenumber::MTN , $this->phone->getOperator('652236987') );
        $this->assertEquals(Phonenumber::MTN , $this->phone->getOperator('653236987') );
        $this->assertEquals(Phonenumber::MTN , $this->phone->getOperator('654236987') );
    }
    /** @test  */
    public function phone_with_prefix_orange()
    {
        $this->assertEquals(Phonenumber::ORANGE , $this->phone->getOperator('237691131446') );
        $this->assertEquals(Phonenumber::ORANGE , $this->phone->getOperator('237655236987') );
        $this->assertEquals(Phonenumber::ORANGE , $this->phone->getOperator('237656236987') );
    }
    /** @test */
    public function phone_without_prefix()
    {
        $this->assertEquals($this->phone->getPhoneWithoutPrefix('237691131446'), '691131446');
    }
    
    
}

src/PhoneFaker.php
<?php

namespace Gildastema\Phonenumber;

class PhoneFaker
{
    /**
     * @param $operator
     * @return string
     * @throws \Exception
     */
    public function generate($operator = null, $prefix=null) : string
    {
        if ($operator === Phonenumber::MTN){
            $phone = $this->randomElement(['67', '68', '650', '651', '652', '653', '654']);

        }elseif ($operator === Phonenumber::ORANGE){
            $phone = $this->randomElement(['69', '655', '656', '657', '658', '659']);
        }else{
            $phone = $this->randomElement(['69', '655', '656', '657', '658', '659', '67', '68', '650', '651', '652', '653', '654']);
        }

        $p = $prefix ?? '';
        if (strlen($phone) === 3){
            return  $p.$phone.random_int(100000,999999);
        }else{
            return $p.$phone.random_int(1000000,9999999);
        }
    }

    private function  randomElement(array $data)
    {
        $len = sizeof($data);
        return $data[ random_int(0, $len -1) ];
    }
}

src/Phonenumber.php
<?php

namespace Gildastema\Phonenumber;

class Phonenumber
{
    const ORANGE = 10;
    const MTN = 11;
    const NEXTTEL = 12;
    const CAMTEL = 13;
    const UNKOWN = -1;
    /**
     * get operator from phone
     *
     * @param string $phone
     * @return integer
     */
    public  function getOperator(string $phone): int
    {
        $phone = self::getPhoneWithoutPrefix(trim($phone));

        if(strlen($phone) !== 9) return self::UNKOWN;
        if ($this->startsWith($phone, '69')) {
            return self::ORANGE;
        }
        if (
            $this->startsWith($phone, '655') || $this->startsWith($phone, '656') || $this->startsWith($phone, '657')
            || $this->startsWith($phone, '658') || $this->startsWith($phone, '659')
        ) {
            return self::ORANGE;
        }
        if ($this->startsWith($phone, '67'))
            return self::MTN;
        if (
            $this->startsWith($phone, '650') || $this->startsWith($phone, '651') || $this->startsWith($phone, '652')
            || $this->startsWith($phone, '653') || $this->startsWith($phone, '654') || $this->startsWith($phone, '680') || $this->startsWith($phone, '681')
            || $this->startsWith($phone, '682')
        ) {
            return self::MTN;
        }
        if ($this->startsWith($phone, '66'))
            return self::NEXTTEL;
        return self::UNKOWN;
    }
    /**
     * Undocumented function
     *
     * @param string $phone
     * @return string
     */
    public  function getPhoneWithoutPrefix(string $phone): string
    {
        if (strlen($phone) == 12) {
            return substr($phone, 3,9);
        }
        if (strlen($phone) == 11) {
            return '6' . str_replace('237', '', $phone);
        }
        if (strlen($phone) == 9) {
            return $phone;
        }
        if (strlen($phone) == 8)
            return '6' . $phone;
        return $phone;
    }


    private function startsWith($string, $startString)
    {
        $len = strlen($startString);
        return (substr($string, 0, $len) === $startString);
    }
}
